File: docs/calculator-engine.js
/**
 * CalculatorEngine
 * -----------------
 * This is the JavaScript twin of your C++ CalculatorEngine class.
 * Notice the architecture is IDENTICAL: pure math, zero knowledge of
 * buttons or the DOM. This file could be reused in a Node.js backend,
 * a different web page, anywhere — exactly like the C++ version could
 * be reused in a CLI tool.
 *
 * In C++, invalid operations throw std::invalid_argument.
 * In JavaScript, the equivalent is throwing a standard Error object.
 * The calling code (app.js) catches it with try/catch — same pattern,
 * different syntax.
 */
class CalculatorEngine {
    add(a, b) {
        return a + b;
    }

    subtract(a, b) {
        return a - b;
    }

    multiply(a, b) {
        return a * b;
    }

    divide(a, b) {
        if (b === 0) {
            throw new Error("Division by zero is undefined.");
        }
        return a / b;
    }

    // The 'op' parameter is a string ("add", "subtract", etc.) instead
    // of a C++ enum class — JavaScript doesn't have true enums, so a
    // string is the idiomatic equivalent here.
    calculate(a, b, op) {
        switch (op) {
            case "add": return this.add(a, b);
            case "subtract": return this.subtract(a, b);
            case "multiply": return this.multiply(a, b);
            case "divide": return this.divide(a, b);
            default: throw new Error("Unknown operation.");
        }
    }

    squareRoot(a) {
        if (a < 0) {
            throw new Error("Cannot take the square root of a negative number.");
        }
        return Math.sqrt(a);
    }

    square(a) {
        return a * a;
    }

    logarithm(a) {
        if (a <= 0) {
            throw new Error("Logarithm is undefined for zero or negative numbers.");
        }
        return Math.log10(a);
    }
}

File: src/CalculatorEngine.cpp
#include "CalculatorEngine.h"
#include <cmath>

CalculatorEngine::CalculatorEngine(QObject *parent)
    : QObject(parent)
{
}

double CalculatorEngine::add(double a, double b) const { return a + b; }
double CalculatorEngine::subtract(double a, double b) const { return a - b; }
double CalculatorEngine::multiply(double a, double b) const { return a * b; }

double CalculatorEngine::divide(double a, double b) const
{
    if (b == 0.0) {
        throw std::invalid_argument("Division by zero is undefined.");
    }
    return a / b;
}

double CalculatorEngine::calculate(double a, double b, Operation op) const
{
    switch (op) {
        case Operation::Add:      return add(a, b);
        case Operation::Subtract: return subtract(a, b);
        case Operation::Multiply: return multiply(a, b);
        case Operation::Divide:   return divide(a, b);
    }
    throw std::logic_error("Unknown operation.");
}

double CalculatorEngine::squareRoot(double a) const
{
    if (a < 0.0) {
        // Real numbers only — no complex number support, so we fail
        // loudly rather than returning NaN silently.
        throw std::invalid_argument("Cannot take the square root of a negative number.");
    }
    return std::sqrt(a);
}

double CalculatorEngine::square(double a) const
{
    return a * a;
}

double CalculatorEngine::logarithm(double a) const
{
    if (a <= 0.0) {
        throw std::invalid_argument("Logarithm is undefined for zero or negative numbers.");
    }
    return std::log10(a);
}